README.md
# `dyn`amic `str`ings

_A C library to safely yet efficiently work with UTF-8–encoded, growable dynamic strings._

- [Code of Conduct](https://github.com/SFM61319/ds/blob/main/CODE_OF_CONDUCT.md)
- [Source Code](https://github.com/SFM61319/ds)
- [Documentation](https://sfm61319.github.io/ds/)
- [Getting Support](https://github.com/SFM61319/ds/blob/main/SUPPORT.md)
- [Reporting Security Vulnerabilities](https://github.com/SFM61319/ds/security/policy)
- [Contributing](https://github.com/SFM61319/ds/blob/main/CONTRIBUTING.md)
- [Changelog](https://github.com/SFM61319/ds/blob/main/CHANGELOG.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/ "Keep a Changelog"), and
this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html "SemVer").

## [Unreleased]

## [`v0.10.0`] - 2023-08-31

### Added

- The `String` type and its related constants and functions.

### Changed

- Change immutable C-string (`ds_cstr_t`) arguments to mutable C-string (`ds_cstr_mut_t`) arguments in all C-string functions.

## [`v0.9.0`] - 2023-08-23

### Added

- `DS_HELPERS_IN_RANGE` and `DS_HELPERS_IN_RANGE_INCLUSIVE` helper macro functions.
- C-string allocator functions.

### Changed

- Make the global memory allocator mutable to support custom memory allocators.
- Replaced `ds_char_in_range` with `DS_HELPERS_IN_RANGE_INCLUSIVE`.

### Removed

- `ds_char_in_range` function in favor of the `DS_HELPERS_IN_RANGE_INCLUSIVE` macro function.

## [`v0.8.0`] - 2023-08-10

### Added

- The `Allocator` type and its related constants and functions.
- A global allocator based on C stdlib's `malloc`, `realloc`, and `free`.

## [`v0.7.0`] - 2023-08-03

### Added

- `Err` `Result` propagator.
- `DS_RESULT_ERR_PTR_IS_NULL` propagator for disallowed `NULL` pointers.

## [`v0.6.0`] - 2023-08-01

### Added

- The `Result` type and its related constants and functions.

## [`v0.5.0`] - 2023-07-31

### Added

- C-string types and their related constants and functions.
- More named character constants.
- A character predicate function pointer type.

## [`v0.4.0`] - 2023-07-27

### Added

- Additional basic `char` functions.

## [`v0.3.0`] - 2023-07-26

### Added

- The `char` type and its related constants and functions.

## [`v0.2.0`] - 2023-07-25

### Added

- Helper constants and functions.
- Usage examples to `usize`-related documentation.

## [`v0.1.0`] - 2023-07-25

### Added

- The `usize` type and its related constants and functions.
- Configuration files and other meta project files.
- CMake for building and testing.

[unreleased]: https://github.com/SFM61319/ds/compare/v0.10.0...HEAD
[`v0.1.0`]: https://github.com/SFM61319/ds/releases/tag/v0.1.0
[`v0.2.0`]: https://github.com/SFM61319/ds/compare/v0.1.0...v0.2.0
[`v0.3.0`]: https://github.com/SFM61319/ds/compare/v0.2.0...v0.3.0
[`v0.4.0`]: https://github.com/SFM61319/ds/compare/v0.3.0...v0.4.0
[`v0.5.0`]: https://github.com/SFM61319/ds/compare/v0.4.0...v0.5.0
[`v0.6.0`]: https://github.com/SFM61319/ds/compare/v0.5.0...v0.6.0
[`v0.7.0`]: https://github.com/SFM61319/ds/compare/v0.6.0...v0.7.0
[`v0.8.0`]: https://github.com/SFM61319/ds/compare/v0.7.0...v0.8.0
[`v0.9.0`]: https://github.com/SFM61319/ds/compare/v0.8.0...v0.9.0
[`v0.10.0`]: https://github.com/SFM61319/ds/compare/v0.9.0...v0.10.0


include/ds/cstr.h
#pragma once

#ifndef DS_CSTR_H
#define DS_CSTR_H 1

//! @file ds/cstr.h
//! @author Avinash Maddikonda (svasssakavi@gmail.com)
//! @brief Declaration of the C-string types, related constants and functions.
//! @since 0.5.0
//! @date 2023-07-31

#include "ds/char.h"
#include "ds/result.h"
#include "ds/usize.h"

#ifdef __cplusplus
extern "C"
{
#endif /* __cplusplus */

  /// @brief An immutable C-string type.
  ///
  /// The @ref ds_cstr_t "immutable C-string" type represents a pointer to an
  /// immutable @ref ds_char_t "char" array.
  ///
  /// @see https://en.wikipedia.org/wiki/Immutable_object
  /// @see https://en.wikipedia.org/wiki/C_string_handling
  typedef ds_char_t const *ds_cstr_t;

  /// @brief A mutable C-string type.
  ///
  /// The @ref ds_cstr_mut_t "mutable C-string" type represents a pointer to a
  /// mutable @ref ds_char_t "char" array.
  ///
  /// @see https://en.wikipedia.org/wiki/Immutable_object
  /// @see https://en.wikipedia.org/wiki/C_string_handling
  typedef ds_char_t *ds_cstr_mut_t;

  /// @brief Attempts to allocate a C-string.
  ///
  /// @warning Always pair a @ref ds_cstr_allocate call with a @ref
  /// ds_cstr_reallocate or @ref ds_cstr_deallocate call (passing the same
  /// pointer). Failing to do so ***will*** lead to memory leaks.
  ///
  /// @param[out] self The allocated C-string on success.
  /// @param size The size of the C-string to allocate.
  ///
  /// @return @ref DS_RESULT_OK "Ok" on success.
  /// @return @ref DS_RESULT_ERR_PTR_IS_NULL "Err(PtrIsNull)" if @p self is
  /// `NULL`.
  /// @return @ref DS_RESULT_ERR_MEM_ALLOC_FAILED "Err(MemAllocFailed)" if a
  /// memory allocation fails.
  extern ds_result_t ds_cstr_allocate (ds_cstr_mut_t *self, ds_usize_t size);

  /// @brief Attempts to reallocate (extend, shrink) the C-string.
  ///
  /// @warning Always pair a @ref ds_cstr_reallocate call with a @ref
  /// ds_cstr_reallocate or @ref ds_cstr_deallocate call (passing the same
  /// pointer). Failing to do so ***will** lead to memory leaks.
  ///
  /// @param src_cstr_ptr A reference to the C-string to reallocate.
  /// @param src_size The old size of the C-string to reallocate.
  /// @param[out] dst_cstr_ptr A reference to the reallocated C-string.
  /// @param dst_size The new size of the reallocated C-string.
  ///
  /// @return @ref DS_RESULT_OK "Ok" on success.
  /// @return @ref DS_RESULT_ERR_PTR_IS_NULL "Err(PtrIsNull)" if @p
  /// src_cstr_ptr or @p dst_cstr_ptr is `NULL`.
  /// @return @ref DS_RESULT_ERR_MEM_ALLOC_FAILED "Err(MemAllocFailed)" if a
  /// memory allocation fails.
  extern ds_result_t ds_cstr_reallocate (ds_cstr_mut_t *src_cstr_ptr,
                                         ds_usize_t src_size,
                                         ds_cstr_mut_t *dst_cstr_ptr,
                                         ds_usize_t dst_size);

  /// @brief Deallocates the C-string referenced by @p self.
  ///
  /// @warning Only pass pointers allocated using @ref ds_cstr_allocate or @ref
  /// ds_cstr_reallocate. Passing other pointers may lead to undefined
  /// behavior.
  ///
  /// @param[out] self A reference to the C-string to deallocate.
  /// @return @ref DS_RESULT_OK "Ok" on success.
  /// @return @ref DS_RESULT_ERR_PTR_IS_NULL "Err(PtrIsNull)" if @p self is
  /// `NULL`.
  extern ds_result_t ds_cstr_deallocate (ds_cstr_mut_t *self);

/// @brief The size of C-strings in bytes.
///
/// ### Examples
///
/// Basic usage:
///
/// ```c
/// assert (DS_CSTR_BYTES == sizeof (ds_char_t *));
/// ```
#define DS_CSTR_BYTES ((ds_usize_t)sizeof (ds_cstr_t))

/// @internal
/// @brief Internal implementation of @ref DS_SCN_CSTR.
#define DS_SCN_CSTR_INTERNAL_IMPL(len) #len "[^\n]%*" DS_SCN_CHAR

/// @brief C-string notation. Can be used in `scanf` to parse and read
/// `len + 1` (@p len = `size - 1`) sized C-string values.
///
/// @param len The length of the C-string (excluding the null-terminator).
/// @warning @p len must be (a macro constant containing) an unsigned integer
/// literal (with no suffixes or preceding zeroes). Any other kind of value,
/// expression, or statement is strictly disallowed.
///
/// @return C-string notation for the given string length @p len.
///
/// @warning It is strongly advised to use
/// [`fgets`](https://en.cppreference.com/w/c/io/fgets "fgets -
/// cppreference.com") instead of `scanf` to read input.
///
/// ### Examples
///
/// Basic usage:
///
/// ```c
/// #define SIZE 15
/// char x[SIZE + 1];
/// scanf("%" DS_SCN_CSTR(SIZE), x);
/// ```
#define DS_SCN_CSTR(len) DS_SCN_CSTR_INTERNAL_IMPL (len)

/// @brief C-string notation. Can be used in `printf` to format and write
/// C-string values.
///
/// ### Examples
///
/// Basic usage:
///
/// ```c
/// ds_cstr_t const x = "abc";
/// printf ("%" DS_PRI_CSTR, x);
/// ```
#define DS_PRI_CSTR "s"

#ifdef __cplusplus
}
#endif /* __cplusplus */

#endif /* DS_CSTR_H */
